rfcs/0077-iceberg-sink-shuffle.md

```diff
@@ -0,0 +1,197 @@
+CREATE SINK s0 AS SELECT * FROM t0 WITH (
+    connector = 'iceberg'
+);
+```
+
+In this case we need to shuffle by `(bucket(id), years(ts))` so that rows with same partition value will go to same actor.
+
+### Case 3: Upsert table with range only partition
+
+Following iceberg table definition has only range partition:
+```sql
+## Iceberg
+CREATE TABLE prod.db.sample (
+    id bigint,
+    data string,
+    category string,
+    ts timestamp)
+USING iceberg
+PARTITIONED BY (years(ts))
+```
+
+And we want to sink following query to iceberg:
+```sql
+CREATE TABLE t0 (id bigint primary key, data string, category string, ts timestamp);
+
+CREATE SINK s0 AS SELECT * FROM t0 WITH (
+    connector = 'iceberg'
+);
+```
+
+In this case we need to shuffle by `(years(ts), id)` to avoid all traffic go to same actor. 
+
+
+### Case 4: Upsert table with not range only partition
+
+Following iceberg table definition has only range partition:
+```sql
+## Iceberg
+CREATE TABLE prod.db.sample (
+    id bigint,
+    data string,
+    category string,
+    ts timestamp)
+USING iceberg
+PARTITIONED BY (bucket(data, 16), years(ts))
+```
+
+And we want to sink following query to iceberg:
+```sql
+CREATE TABLE t0 (id bigint primary key, data string, category string, ts timestamp);
+
+CREATE SINK s0 AS SELECT * FROM t0 WITH (
+    connector = 'iceberg'
+);
+```
+
+In this case we need to shuffle by `(bucket(data), years(ts))` so that rows with same will go to same actor. Notice that in iceberg we don't need to shuffle by `stream_key`, e.g. `id` since it only requires that insert/delete of same row will go to same partition. 
+
+### Implementation
+
+There are two possible implementations for shuffling by iceberg partition columns:
+
+1. Add a new `IcebergPartitionExecutor`, which calculates the partition value of each record and adds it to the record, then asking the dispatcher executor to do hash shuffle according to the partition value. The plan is like following:
+
+```mermaid
+flowchart TD
+    A[Source] --> B[IcebergPartitionExecutor]
+    B --> C[DispatcherExecutor]
+    C -->|"partition_value"| D[IcebergSink]
+```
+
+The `IcebergPartitionExecutor` will be a `StreamExecutor`, which calculates the partition value of each record and adds it to the record. The `DispatcherExecutor` doesn't need to change much, and do hash shuffle according to the partition value.
+
+2. Extend dispatcher executor to support iceberg partition shuffle. The plan is like following:
+
+```mermaid 
+flowchart TD
+    A[Source] --> B[DispatcherExecutor]
+    B -->|IcebergDispatcher| C[IcebergSink]
+```
+
+In this approach we need to add an extra `IcebergDispatcher` to dispatcher executor. The `IcebergDispatcher` will calculate the partition value of each record and do hash shuffle according to the partition value.
+
+I prefer approach 1 since it's more extensible and does not change too much current shuffle implementation, e.g. other lakehouse sinks (delta lake) could have similar approach.
```

Comment:
Hmm, actually I would prefer approach 2 because we consider the `Merge` and `Dispatch` as very special operators and some implementations depend on this assumption. 

------

Another thing worth to mention is: As RisingWave is the writter of downstream Iceberg, we can choose to support only a **subset** of Iceberg's partition feature. Generally, Iceberg supports partition *expressions* rather than just a list of columns. However, we may only support a list of columns as partition key, which can simplify our design.

This is also [what Flink did](https://iceberg.apache.org/docs/latest/flink-ddl/):

> PARTITION BY (column1, column2, ...) to configure partitioning, Flink does not yet support hidden partitioning.

Reply:
> Hmm, actually I would prefer approach 2 because we consider the Merge and Dispatch as very special operators and some implementations depend on this assumption.

I still prefer approach 1 for several reasons:
1. `Dispatch` executor in fact calls `Dispatcher` for different  distributions in risingwave, and I don't want to introduce another distribution which serves only external system.
2. It wastes computed partition values.

Reply:
> Another thing worth to mention is: As RisingWave is the writter of downstream Iceberg, we can choose to support only a subset of Iceberg's partition feature. Generally, Iceberg supports partition expressions rather than just a list of columns. However, we may only support a list of columns as partition key, which can simplify our design.

You misunderstood flink's iceberg implementation. The fact is that it doens't support ddl statement, but the writer still respect partition spec created by other engines such as spark, you can try it yourself. 

Also I really don't like the statement that we should not do sth because  flink doesn't support it. Why users should choose use if we blindly follow flink?

Reply:
> Also I really don't like the statement that we should not do sth because flink doesn't support it

No, this is not my motivation. My whole point is:

> As RisingWave is the writter of downstream Iceberg, we can choose to support only a subset of Iceberg's partition feature.

So I'd like to leverage this to simplify our design.

Reply:
I have some misunderstanding. Now it LGTM